src/shared/dns-resolver.c

```diff
@@ -0,0 +1,144 @@
+                offset += 2;
+                if (offset + plen > len)
+                        return -EBADMSG;
+
+                switch (key) {
+                /* Mandatory keys must be understood by the client, otherwise the record should be discarded.
+                 * Automatic mandatory keys must not appear in the mandatory parameter, so these are all
+                 * supplementary. We don't understand any supplementary keys, so if the mandatory parameter
+                 * is present, we cannot use this record.*/
+                case DNS_SVC_PARAM_KEY_MANDATORY:
+                        if (plen > 0)
+                                return -EBADMSG;
+                        break;
+
+                case DNS_SVC_PARAM_KEY_ALPN:
+                        if (plen == 0)
+                                return 0;
+                        alpn = true; /* alpn is required. Record that the requirement is met. */
+
+                        size_t poff = offset;
+                        size_t pend = offset + plen;
+                        while (poff < pend) {
+                                uint8_t alen = option[poff++];
+                                if (poff + alen > len)
+                                        return -EBADMSG;
+                                if (alen == 3 && strneq((const char*) &option[poff], "dot", alen))
+                                        SET_FLAG(transports, SD_DNS_ALPN_DOT, true);
+                                if (alen == 2 && strneq((const char*) &option[poff], "h2", alen))
+                                        SET_FLAG(transports, SD_DNS_ALPN_HTTP_2_TLS, true);
+                                if (alen == 2 && strneq((const char*) &option[poff], "h3", alen))
+                                        SET_FLAG(transports, SD_DNS_ALPN_HTTP_3, true);
+                                if (alen == 3 && strneq((const char*) &option[poff], "doq", alen))
+                                        SET_FLAG(transports, SD_DNS_ALPN_DOQ, true);
+                                poff += alen;
+                        }
+                        if (poff != pend)
+                                return -EBADMSG;
+                        break;
+
+                case DNS_SVC_PARAM_KEY_PORT:
+                        if (plen != 2)
+                                return -EBADMSG;
+                        port = unaligned_read_be16(&option[offset]);
+                        /* Server should indicate default port by omitting this param */
+                        if (port == 0)
+                                return -EBADMSG;
+                        break;
+
+                /* RFC9463 § 5.1 service params MUST NOT include ipv4hint/ipv6hint */
+                case DNS_SVC_PARAM_KEY_IPV4HINT:
+                case DNS_SVC_PARAM_KEY_IPV6HINT:
+                        return -EBADMSG;
+
+                case DNS_SVC_PARAM_KEY_DOHPATH: {
+                        int r;
```

Comment:
`int r` is special to us, we always define it at outermost function scope. hence please move this up